### DISCUSSION TITLE
How to check if proxy is set

### DISCUSSION THREAD

**Discussion by forthrin** [HoneyryderChuck/httpx#17] [Q&A]
How does one check if a proxy is set on a HTTPX object?

(More generally, is there any correlation between elements in `pp` output and accessing those elements programmatically?)

```
#<#<Class:0x00000000babeface>:0x00000000babeface
 @options=
  #<#<Class:0x00000000babeface>:0x00000000babeface
  (...)
   @proxy={:uri=>"socks5://127.0.0.1:1234"},
```

**Comment by HoneyryderChuck** [comment]
> How does one check if a proxy is set on a HTTPX object?

Why do you need to know it, i.e. what's the use-case?

A session is usually manually set by the developer, so you should already know whether a session sets up a proxy. The exception is when your system sets up the `http_proxy` variables, where all sessions will then have proxy set up.

Having a proxy set up does also not mean all requests are proxied (there are `no_proxy` rules, system or manual, and also 305 status codes).

A very barebones way to answer "is the proxy plugin" loaded is `session.class.default_options.respond_to?(:proxy)` though.

It would help if you could describe your use-case, to understand whether this is worth exposing via a friendlier API.

**Comment by forthrin** [comment]
The use case is having a safety net:

- To absolutely, positively make sure that the proxy is in place before requests are run

In the case that:

- The proxy may be applied in some general setup code "far, far away" from the caller

And henceforth, in the case that this code may be:

- Temporarily commented out by some doofus (including self)
- Failing to run for some other reason

**Reply by HoneyryderChuck** [reply]
I don't see the value of supporting such a feature. As per [the docs](https://honeyryderchuck.gitlab.io/httpx/wiki/Proxy.html), by using `.plugin(:proxy)` or the env vars mentioned, requests will be proxied unless in the documented exceptions. 

However, you can use debugging logs to make sure the request is proxied, while you test it. Or you can build something for yourself via a custom plugin. And you can yourself inspect the changelog and test coverage to ensure that no doofus (please don't call me that again) broke proxy plugin tests.

**Comment by forthrin** [comment]
First, "yourself" was reflecting on the undersigned, or whoever uses library. Updated to "self" as to prevent further tension.

So, in conclusion, there is no shorthand way to read the value (IP address) of the proxy (or nil if it doesn't exist)?

Only plausible way is to use the suggested `session.class.default_options.respond_to?(:proxy)`?

**Reply by HoneyryderChuck** [reply]
> So, in conclusion, there is no shorthand way to read the value (IP address) of the proxy (or nil if it doesn't exist)?

Do you want to know the proxy IP, or do you want to access the peer address to which a request was sent? The first one should be your responsibility (as the owner of the machine and environment where the software is deployed), whereas the second is a more reasonable feature request (currently, `httpx` does not provide it).

> Only plausible way is to use the suggested session.class.default_options.respond_to?(:proxy)?

If you want to use "vanilla" API, yes. However, I'd recommend you to write a custom plugin to detect these things yourself. Feel free to get hints from the docs and [this plugin used in the proxy tests](https://gitlab.com/os85/httpx/-/blob/master/test/support/proxy_response_detector.rb).
